### PR TITLE
[interp] fixes around delegates, icalls and resolving types

### PR DESCRIPTION
**Pull request by lewurm**
@vargaz please have a look at the general mini changes

### CODE REVIEW

**Comment by lewurm** [comment]
@monojenkins build interpreter

**Comment by lewurm** [comment]
@monojenkins build interpreter

File: mono/mini/interp/transform.c [hunk 1]
```diff
@@ -2966,6 +2984,22 @@ generate (MonoMethod *method, RuntimeMethod *rtm, unsigned char *is_bb_start, Mo
...
+				g_assert (key < TLS_KEY_NUM);
+				ADD_CODE (&td, MINT_MONO_TLS);
+				WRITE32 (&td, &key);
+				PUSH_SIMPLE_TYPE (&td, STACK_TYPE_MP);
+				break;
+			}
+			case CEE_MONO_ATOMIC_STORE_I4:
+				CHECK_STACK (&td, 2);
+				SIMPLE_OP (td, MINT_STIND_I4);
+				ADD_CODE (&td, MINT_MONO_MEMORY_BARRIER);
```
Comment (alexrp) [inline mono/mini/interp/transform.c]: This implementation is not quite correct. The strongest barrier this instruction supports, `MONO_MEMORY_BARRIER_SEQ`, would require a full barrier on both sides of the store.

Additionally, doing the store itself as regular C `*dst = src`, as opposed to using an atomic intrinsic, is prone to esoteric compiler optimizations even with memory barriers. The best thing to do here would be to just call `InterlockedWrite` from `utils/atomic.h` to implement this instruction.

**Comment by lewurm** [comment]
@monojenkins build interpreter

File: mono/mini/interp/transform.c (hunk 1 above)
Comment (lewurm) [inline mono/mini/interp/transform.c]: thanks @alexrp! I updated the PR.

File: mono/mini/interp/interp.c [hunk 2]
```diff
@@ -1903,9 +1903,13 @@ mono_interp_create_method_pointer (MonoMethod *method, MonoError *error)
 	MonoMethod *wrapper;
 	RuntimeMethod *rmethod;
 
-	if (method->wrapper_type && method->wrapper_type != MONO_WRAPPER_RUNTIME_INVOKE)
+	if (method->wrapper_type && method->wrapper_type == MONO_WRAPPER_NATIVE_TO_MANAGED)
 		return NULL;
 
+	/* HACK: method_ptr of delegate should point to a runtime method*/
+	if (method->wrapper_type && method->wrapper_type == MONO_WRAPPER_DYNAMIC_METHOD)
+		return mono_interp_get_runtime_method (mono_domain_get (), method, error);
```
Comment (vargaz) [inline mono/mini/interp/interp.c]: This is the same line as the line below.

File: mono/mini/interp/transform.c [hunk 3]
```diff
@@ -916,6 +916,21 @@ interp_transform_call (TransformData *td, MonoMethod *method, MonoMethod *target
 	}
 }
 
+static MonoClassField *interp_field_from_token (MonoMethod *method, guint32 token, MonoClass **klass, MonoGenericContext *generic_context)
+{
```
Comment (vargaz) [inline mono/mini/interp/transform.c]: The function name should be on a new line.

**Comment by vargaz** [comment]
Looks ok.

File: mono/mini/interp/transform.c (hunk 3 above)
Comment (lewurm) [inline mono/mini/interp/transform.c]: done.

**Pull request merged**